File: Python/Algorithms/1315.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumEvenGrandparent(self, root) -> int:
        def traverse_sum(node, parent=None):
            sum_ = 0
            val, left, right = node.val, node.left, node.right
            if left:
                if parent is not None and parent & 1 == 0:
                    sum_ = sum_ + left.val
                sum_ = sum_ + traverse_sum(left, val)
            if right:
                if parent is not None and parent & 1 == 0:
                    sum_ = sum_ + right.val
                sum_ = sum_ + traverse_sum(right, val)
            return sum_

        # we know min nodes == 1.
        return traverse_sum(root)

File: C/Algorithms/1315.c
// Definition for a binary tree node.
struct TreeNode {
  int val;
  struct TreeNode *left;
  struct TreeNode *right;
};

typedef struct TreeNode Node;

int traverse_sum(Node *node, int parent) {
  int sum = 0, val = node->val;
  Node *left = node->left, *right = node->right;

  // If left, add left val to sum and then add
  // whatever sum we get from recursing on left.
  // Similarly for right.
  if (left) {
    if ((parent & 1) == 0)
      sum += left->val;
    sum += traverse_sum(left, val);
  }
  if (right) {
    if ((parent & 1) == 0)
      sum += right->val;
    sum += traverse_sum(right, val);
  }
  return sum;
}

int sumEvenGrandparent(Node *root) {
  int val = root->val, sum = 0;
  Node *left = root->left, *right = root->right;
  // Call directly on children of root so we can
  // pass root val.
  if (left)
    sum += traverse_sum(left, val);
  if (right)
    sum += traverse_sum(right, val);
  return sum;
}
